src/rgw/rgw_notify.cc

```diff
@@ -2,20 +2,530 @@
+          << " ok" <<  dendl;
+        if (perfcounter) perfcounter->inc(l_rgw_pubsub_push_ok);
+        return true;
+      }
+    } catch (const RGWPubSubEndpoint::configuration_error& e) {
+      ldout(cct, 5) << "WARNING: failed to create push endpoint: " 
+          << record_with_endpoint.push_endpoint << " for entry: " << entry.marker << ". error: " << e.what() << " (will retry) " << dendl;
+      return false;
+    }
+  }
+
+  // processing of a specific queue
+  void process_queue(const std::string& queue_name, const bool& owned, spawn::yield_context yield) {
+    constexpr auto max_elements = 1024;
+    auto is_idle = false;
+    const std::string start_marker;
+    while (owned) {
+      if (is_idle) {
+        Timer timer(io_context);
+        timer.expires_from_now(std::chrono::microseconds(queue_idle_sleep_us));
+        boost::system::error_code ec;
+	      timer.async_wait(yield[ec]);
+      }
+      is_idle = true;
+
+      bool truncated = false;
+      std::string end_marker;
+      std::vector<cls_queue_entry> entries;
+      auto total_entries = 0U;
+
+      {
+        librados::ObjectReadOperation op;
+        bufferlist obl;
+        int rval;
+        cls_2pc_queue_list_entries(op, start_marker, max_elements, &obl, &rval);
+        auto ret = rgw_rados_operate(rados_ioctx, queue_name, &op, nullptr, optional_yield(io_context, yield));
+        if (ret == -ENOENT) {
+          // queue was deleted
+          ldout(cct, 5) << "INFO: queue: " 
+            << queue_name << ". was removed. processing will stop" << dendl;
+          return;
+        }
+        if (ret < 0) {
+          ldout(cct, 5) << "WARNING: failed to get list of entries in queue: " 
+            << queue_name << ". error: " << ret << " (will retry)" << dendl;
+          continue;
+        }
+        ret = cls_2pc_queue_list_entries_result(obl, entries, &truncated, end_marker);
+        if (ret < 0) {
+          ldout(cct, 5) << "WARNING: failed to parse list of entries in queue: " 
+            << queue_name << ". error: " << ret << " (will retry)" << dendl;
+          continue;
+        }
+      }
+      total_entries = entries.size();
+      if (total_entries == 0) {
+        // nothing in the queue
+        continue;
+      }
+      // log when queue is not idle
+      ldout(cct, 20) << "INFO: found: " << total_entries << " entries in: " << queue_name <<
+        ". end marker is: " << end_marker << dendl;
+      
+      is_idle = false;
+      auto has_error = false;
+      auto remove_entries = false;
+      auto entry_idx = 1U;
+      tokens_waiter waiter(io_context);
+      for (auto& entry : entries) {
+        if (has_error) {
+          // bail out on first error
+          break;
+        }
+        // TODO pass entry pointer instead of by-value
+        spawn::spawn(yield, [this, &queue_name, entry_idx, total_entries, &end_marker, &remove_entries, &has_error, &waiter, entry](spawn::yield_context yield) {
```

Comment:
we want to process other entries while we wait for ack/nack for an entry.

Reply:
i think it was a question about performance impact: each call to spawn() will allocate 2 shared_ptrs for coroutine state, along with a 128k coroutine stack

i think the `push_endpoint->send_to_completion_async()` interface relies on having a yield context, but it might be worth exploring a different async interface that avoids spawning a coroutine for each entry. ideally that interface would support the `CompletionToken` from asio's [extensible asynchronous model](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3896.pdf) so we could just pass a light-weight function object for the completion - but `send_to_completion_async()` is a virtual interface, so can't be templated

Reply:
@cbodley that makes sense. I will look into that, but not in the scope of this PR